notebooks/decorators_nb.py

```diff
@@ -1,2 +1,223 @@
 #%%
-# to be done
+# according to this tutorial: https://realpython.com/primer-on-python-decorators/
+import functools
+
+
+def my_first_decorator(func):
+    def wrapper():
+        print("An event before the function is called")
+        func()
+        print("An event after the function is called")
+    return wrapper
+
+
+def say_whee():
+    print("Whee!")
+
+
+say_whee = my_first_decorator(say_whee)
+
+say_whee()
+
+
+@my_first_decorator
+def say_hey():
+    print("Hey")
+
+
+say_hey()
+
+# thus, decorators wrap functions, modifying their behavior, but not themselves
+
+#%%
+# a calculator from a video course, once again revisited.
+# trying a decorator for catching invalid input
+import functools
+
+
+def invalid_input_catch(func):
+    @functools.wraps(func)
+    def wrapper():
+        try:
+            return func()  # if we want decorator to return a value, use return keyword
+        except ValueError:
+            print("Invalid input: you should enter a number")
+            return wrapper()  # if not a number, try again
+    return wrapper
+
+
+@invalid_input_catch
+def get_num():
+    num = float(input("Please enter a number"))
+    return num
+
+
+num1 = get_num()
+
+op = input("Please specify the operation: +, -, *, / or **")
+while op not in ["+", "-", "*", "/", "**"]:
+    print("You've entered a wrong operator, please retry")
+    op = input("Please specify the operation: +, -, *, / or **")
+
+num2 = get_num()
+
+
+def exponent(base, power):
+    exp_result = 1
+    for i in range(power):
+        exp_result *= base
+    return exp_result
+
+
+if op == "+":
```

Comment:
this should be a method

Reply:
I've turned it to a static method. It will do?